proposals/0094-deprecate-executable-update.md

```diff
@@ -0,0 +1,67 @@
+---
+simd: '0094'
+title: Deprecate executable update in bpf loader
+authors:
+  - Haoran Yi
+category: Standard
+type: Core
+status: Draft
+created: 2023-12-13
+feature: (fill in with feature tracking issues once accepted)
+extends: 0162
+---
+
+## Summary
+
+Deprecate executable update during bpf program deployment.
+
+## Motivation
+
+Currently, when a program is deployed, the "executable" flag on the account is
+set to true for legacy reasons. However, (a) old bpf loaders, such as v1 and v2,
+are disabled on mainnet (SIMD-0093); (b) we are going to remove accounts
+executable checks for runtime execution (SIMD-0162). (c) the future bpf
+loader-v4 no longer need "executable" flag.
+
+Because of these above reasons, "executable" flag becomes irrelevant and
+provides no functional benefit for runtime execution. Therefore, we want to
+deprecate its update during bpf program deployment for loader-v3.
+
+## Alternatives Considered
+
+None
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+When the feature - "deprecate executable update" is activated, the bpf loader-V3
+will no longer update *executable* flag to true after program deployment.
+
+The "executable" flag remains false after program deployment.
+
+## Impact
+
+This will affect the following scenarios.
+
+- program accounts hash after deployment. Since this change is guarded by a
+  feature, program accounts hash change won't break consensus.
+- any dapps that depend on `is_executable` to be true on the serialized
+  instruction accounts. Existing program deployed on chain will work fine.
+  However, if the program is redeployed, it may be broken. Before redeployment,
+  dapps developer will need to check and update the program if it depends on
+  `is_executable`.
```

Comment:
Possibly a big deal for dApp developers, cc @acheroncrypto

Reply:
> Possibly a big deal for dApp developers, cc @acheroncrypto

It shouldn't affect Anchor as long as `AccountInfo.executable` doesn't change (see https://github.com/solana-labs/solana/issues/34425#issuecomment-1915250031).

Reply:
The usage of `is_executable` and `executable` is a bit confusing. I think using something like `<TYPE>.is_executable` would help resolve this ambiguity.

Reply:
> It shouldn't affect Anchor as long as `AccountInfo.executable` doesn't change (see [solana-labs/solana#34425 (comment)](https://github.com/solana-labs/solana/issues/34425#issuecomment-1915250031)).

Yeah. updated.

Reply:
> It shouldn't affect Anchor as long as `AccountInfo.executable` doesn't change (see [solana-labs/solana#34425 (comment)](https://github.com/solana-labs/solana/issues/34425#issuecomment-1915250031)).

@acheroncrypto With this proposal, `AccountInfo.executable`  will change. In the original implementation that you mentioned above, `AccountInfo.executable` doesn't change because we emulate it at serialization time. But that turns out to be very expensive - 10X increase of serialization. Therefore,  in this proposal, we won't emulate executable at serialization time. Therefore, dapp can't depends on `AccountInfo.executable` directly.